packages/peregrine/src/Router/resolveUnknownRoute.js

```diff
@@ -63,7 +63,13 @@ function remotelyResolveRoute(opts) {
  * @returns {Promise<{rootChunkID: number, rootModuleID: number}>}
  */
 function tempGetWebpackChunkData(pageType, webpackPublicPath) {
-    return fetch(new URL('roots-manifest.json', webpackPublicPath))
+    // In dev mode, `webpackPublicPath` may be a fully qualified URL.
+    // In production mode, it may be a pathname, which makes it unsafe
+    // to use as an API base. Normalize it as a full path using a DOM node
+    // as a native URL parser.
+    const parser = document.createElement('a');
```

Comment:
Can you elaborate? What does this workaround do that [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) doesn't?

Reply:
It's guaranteed to produce an absolute URL from a possibly-relative URL. The `webpackPublicPath` argument is relative in production and absolute in dev, so we need something that can handle both. URL objects cannot:

```js
// URL constructors deal in absolute URLs.
try {
  new URL('/hermano')
} catch(e) {
  console.warn(e.message, '/hermano'); // Failed to construct 'URL': Invalid URL
}

// They can use relative URLs if you supply a "base URL" as a second argument.
const absoluteBase = 'https://github.com/magento-research/pwa-studio/';
const url = new URL('pulls/165', absoluteBase);
console.log(url)
// Url { protocol: "https:" hostname: "github.com", pathname: "/magento-research/pwa-studio/pulls/165" }

// If the base URL is relative, they throw.
try {
  new URL('pulls/165', '/magento-research/pwa-studio/');
} catch(e) {
  console.warn(e.message, '/magento-research/pwa-studio/');
}

// Anchor tags will accept absolute or relative 'href' attributes:
const parserTag = document.createElement('a');
parserTag.setAttribute('href','/magento-research/pwa-studio');
// But their properties are always absolute:
console.log('parserTag.getAttribute("href")', parserTag.getAttribute("href"));
console.log('parserTag.href', parserTag.href); // "https://github.com/magento-research/pwa-studio/";

// So you can guarantee an absolute URL built with URL semantics
// if you can guarantee an absolute base, which we do with our parser.
// Which is a lot safer than string concatenation.

function appendPathToUrl(path, base) {
   const parser = document.createElement('a');
   parser.setAttribute('href', base);
   return new URL(path, parser.href).toString()
}

console.log(
  'true or false: using an anchor tag as a parser normalizes absolute URLs:'
  appendPathToUrl('pulls/165', '/magento-research/pwa-studio/') === 
  appendPathToUrl('pulls/165', 'https://github.com/magento-research/pwa-studio/')
);
// true
```

Reply:
I see, makes perfect sense to me now. Thanks for the explanation. 👍